src/codegen/build_module.cc

```diff
@@ -0,0 +1,315 @@
+  result << target_name;
+  for (const auto &x : options) {
+    result << " " << x;
+  }
+  return result.str();
+}
+
+Target TargetFromName(const std::string& name) {
+  if (name == "llvm") {
+    return target::llvm();
+  } else if (name == "cuda" || name == "nvptx") {
+    return target::cuda();
+  } else if (name == "rocm" || name == "opencl") {
+    /* For now, assume rocm schedule for opencl */
+    return target::rocm();
+  } else if (name == "metal") {
+    return target::metal();
+  } else if (name == "stackvm" || name == "ext_dev") {
+    return target::stackvm();
+  } else {
+    LOG(ERROR) << "Unknown target name " << name;
+    return target::stackvm();
+  }
+}
+
+bool StartsWith(const std::string& str, const std::string& pattern) {
+  return str.compare(0, pattern.length(), pattern) == 0;
+}
+
+std::string GetDeviceName(const std::string& target_str) {
+  std::istringstream ss(target_str);
+  std::string target_name;
+  ss >> target_name;
+
+  std::string item;
+  while (ss >> item) {
+    if (StartsWith(item, "-device=")) {
+      return item.substr(std::string("-device=").length());
+    }
+  }
+
+  return "";
+}
+
+Target Target::create(const std::string& target_str) {
+  if (target_str.length() == 0) {
+    LOG(ERROR) << "target_str must not be empty";
+  }
+
+  std::istringstream ss(target_str);
+  std::string target_name;
+
+  ss >> target_name;
+  auto device_name = GetDeviceName(target_str);
+
+  auto result = device_name == "rasp" ?
+    target::rasp() :
+    TargetFromName(target_name);
+
+  std::string item;
+  while (ss >> item) {
+    result.options.push_back(item);
+  }
+
+  return result;
+}
+
+namespace target {
+Target llvm() {
+  std::unordered_set<std::string> keys({ "llvm", "cpu" });
+  std::vector<std::string> options;
+  return Target("llvm", kDLCPU, 512, 1, keys, options);
+}
+
+Target cuda() {
+  std::unordered_set<std::string> keys({ "cuda", "gpu" });
+  std::vector<std::string> options;
+  return Target("cuda", kDLGPU, 512, 32, keys, options);
+}
+
+Target rocm() {
+  std::unordered_set<std::string> keys({ "rocm", "gpu" });
+  std::vector<std::string> options;
+  return Target("rocm", kDLROCM, 256, 1, keys, options);
+}
+
+Target metal() {
+  std::unordered_set<std::string> keys({ "gpu" });
+  std::vector<std::string> options;
+  return Target("metal", kDLMetal, 256, 1, keys, options);
+}
+
+Target rasp() {
+  std::unordered_set<std::string> keys({ "llvm", "cpu" });
+  std::vector<std::string> options({
+    "-device=rasp",
+    "-mtriple=armv7l-none-linux-gnueabihf",
+    "-mcpu=cortex-a53",
+    "-mattr=+neon"
+  });
+  return Target("llvm", kDLCPU, 512, 1, keys, options);
+}
+
+Target stackvm() {
+  std::unordered_set<std::string> keys({ "stackvm", "cpu" });
+  std::vector<std::string> options;
+  return Target("stackvm", kDLCPU, 512, 1, keys, options);
+}
+}  // namespace target
+
+bool LLVMEnabled() {
+  const runtime::PackedFunc* pf = runtime::Registry::Get("codegen.llvm_target_enabled");
```

Comment:
This function is not correctly implemented, refer to https://github.com/dmlc/tvm/blob/master/src/codegen/codegen.cc#L24